data/txDupCache.go

```diff
@@ -250,13 +308,26 @@ func (c *txSaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
 		d = &dn
 	}
 
-	c.cur[*d] = struct{}{}
-	return d, false
+	vals = cacheValue{sender: {}}
+	c.cur[*d] = vals
+	vals = copy(vals)
+	return d, vals, false
 }
 
 // DeleteByKey from the cache by using a key used for insertion
 func (c *txSaltedCache) DeleteByKey(d *crypto.Digest) {
-	c.digestCache.Delete(d)
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.cur, *d)
+	delete(c.prev, *d)
+}
+
+// Len returns size of a cache
+func (c *txSaltedCache) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return len(c.cur) + len(c.prev)
 }
```

Comment:
could len be managed proactively as two atomic ints? If we did that, we could avoid taking a lock on len checks I think?